### PR TITLE
Prevent creation of duplicates on bulk indexing with auto generated ids and retry

### LINKED ISSUE #8788: Duplicate id in index
I decided to reindex my data to take advantage of `doc_values`, but one of 30 indices (~120m docs in each) got less documents after reindexing. I reindexed again and docs disappeared again.

Then I bisected the problem to specific docs and found that some docs in source index has duplicate ids.

```
curl -s "http://web245:9200/statistics-20141110/_search?pretty&q=_id:1jC2LxTjTMS1KHCn0Prf1w"
```

```json
{
  "took" : 1156,
  "timed_out" : false,
  "_shards" : {
    "total" : 5,
    "successful" : 5,
    "failed" : 0
  },
  "hits" : {
    "total" : 2,
    "max_score" : 1.0,
    "hits" : [ {
      "_index" : "statistics-20141110",
      "_type" : "events",
      "_id" : "1jC2LxTjTMS1KHCn0Prf1w",
      "_score" : 1.0,
      "_source":{"@timestamp":"2014-11-10T14:30:00+0300","@key":"client_belarussia_msg_sended_from_mutual__22_1","@value":"149"}
    }, {
      "_index" : "statistics-20141110",
      "_type" : "events",
      "_id" : "1jC2LxTjTMS1KHCn0Prf1w",
      "_score" : 1.0,
      "_source":{"@timestamp":"2014-11-10T14:30:00+0300","@key":"client_belarussia_msg_sended_from_mutual__22_1","@value":"149"}
    } ]
  }
}
```

Here are two indices, source and destination:

```
health status index                  pri rep docs.count docs.deleted store.size pri.store.size
green  open   statistics-20141110      5   0  116217042            0     12.3gb         12.3gb
green  open   statistics-20141110-dv   5   1  116216507            0     32.3gb         16.1gb
```

Segments of problematic index:

```
index               shard prirep ip            segment generation docs.count docs.deleted    size size.memory committed searchable version compound
statistics-20141110 0     p      192.168.0.190 _gga         21322   14939669            0   1.6gb     4943008 true      true       4.9.0   false
statistics-20141110 0     p      192.168.0.190 _isc         24348   10913518            0   1.1gb     4101712 true      true       4.9.0   false
statistics-20141110 1     p      192.168.0.245 _7i7          9727    7023269            0   766mb     2264472 true      true       4.9.0   false
statistics-20141110 1     p      192.168.0.245 _i01         23329   14689581            0   1.5gb     4788872 true      true       4.9.0   false
statistics-20141110 2     p      192.168.1.212 _9wx         12849    8995444            0 987.7mb     3326288 true      true       4.9.0   false
statistics-20141110 2     p      192.168.1.212 _il1         24085   13205585            0   1.4gb     4343736 true      true       4.9.0   false
statistics-20141110 3     p      192.168.1.212 _8pc         11280   10046395            0     1gb     4003824 true      true       4.9.0   false
statistics-20141110 3     p      192.168.1.212 _hwt         23213   13226096            0   1.3gb     4287544 true      true       4.9.0   false
statistics-20141110 4     p      192.168.2.88  _91i         11718    8328558            0 909.2mb     2822712 true      true       4.9.0   false
statistics-20141110 4     p      192.168.2.88  _hms         22852   14848927            0   1.5gb     4777472 true      true       4.9.0   false
```

The only thing that happened with index besides indexing is optimizing to 2 segments per shard.

### PR DESCRIPTION
**Pull request by brwe**
If bulk index request fails due to a disconnect, unavailable shard etc, the request is
retried before actually failing. However, even in case of failure the documents
might already be indexed. For autogenerated ids the request must not add the
documents again and therefore canHaveDuplicates must be set to true.

There are two options to fix this in this pull request. 
1. Just add the flag canHaveDuplicates to the request (30fac69). The user then gets a version conflict in the bulk response (see https://github.com/brwe/elasticsearch/commit/30fac69e1a5ae704845094b6e8f8bf194d2b5e0a#diff-38d197a0e6b168e3ca10ca0553b238adR111) which is confusing since the document cannot have been there before if the id was autogenerated in case the user only uses auto generated ids.
2. We additionally prevent the version conflict by checking in InternalEngine for versions and autogenerated ids (9802466, see https://github.com/brwe/elasticsearch/commit/980246615725d2420c56a31171d4a1e93f47b722#diff-5666f4a211c1b7bbaf76addf0c27451bR504). However, this means that if there is an actual version conflict because a document was added before with this id, then the user will not be notified. 

I am unsure which of the two is the lesser evil. 

closes #8788

### CODE REVIEW

**Comment by clintongormley** [comment]
Nice catch!  

> We additionally prevent the version conflict by checking in InternalEngine for versions and autogenerated ids (9802466, see brwe@9802466#diff-5666f4a211c1b7bbaf76addf0c27451bR504). However, this means that if there is an actual version conflict because a document was added before with this id, then the user will not be notified.

In fact, this could also happen with a user-specified ID and version number. If the request succeeds on the first request, then it will fail on a bulk retry with a conflict exception.  I think there is no way around this, as bulk requests are not transactional.

I think the right thing to do here is to report the conflict, and to document the corner-case that this can happen during node restarts.

**Comment by kimchy** [comment]
Actually, in the auto generated id one, it makes more sense to suppress the version conflict exception for the auto generated id case, since in the typical case (no failures), an existing id clashing with the auto generated one will not be reported to begin with (thats part of the optimization we do for auto generated id case).

File: src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import static org.hamcrest.Matchers.greaterThan;
+
+@ElasticsearchIntegrationTest.ClusterScope(scope = ElasticsearchIntegrationTest.Scope.SUITE)
+public class ExceptionRetryTests extends ElasticsearchIntegrationTest {
+
+    volatile boolean exceptionThrown = false;
+
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return ImmutableSettings.builder()
```
Comment (s1monw) [inline src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java]: for master you don't need to specify the gateway.type we only have what used to be local!

File: src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+        //create a transport service that throws a ConnectTransportException for one bulk request and therefore triggers a retry.
+        for (NodeStats dataNode : dataNodeStats) {
+            MockTransportService mockTransportService = ((MockTransportService) internalCluster().getInstance(TransportService.class, dataNode.getNode().name()));
+            mockTransportService.addDelegate(internalCluster().getInstance(Discovery.class, unluckyNode.getNode().name()).localNode(), new MockTransportService.DelegateTransport(mockTransportService.original()) {
+
+                @Override
+                public void sendRequest(DiscoveryNode node, long requestId, String action, TransportRequest request, TransportRequestOptions options) throws IOException, TransportException {
+                    super.sendRequest(node, requestId, action, request, options);
+                    if (action.equals("indices:data/write/bulk[s]") && !exceptionThrown) {
```
Comment (s1monw) [inline src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java]: can this be a constant somehow? I wonder if we have it as a constant already somewhere?

File: src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                        logger.info("Throw disconnected exception");
```
Comment (s1monw) [inline src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java]: can this be a debug log?

File: src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            bulkBuilder.add(client().prepareIndex("index", "type").setSource(doc));
+        }
+
+        BulkResponse response = bulkBuilder.get();
+        if (response.hasFailures()) {
+            for (BulkItemResponse singleIndexRespons : response.getItems()) {
+                if (singleIndexRespons.isFailed()) {
+                    if (singleIndexRespons.getFailure().getStatus() != RestStatus.CONFLICT) {
+                        //This measn that we accept to get a conflict when adding a doucument for the first time which seems odd.
+                        fail();
```
Comment (s1monw) [inline src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java]: can we but some more info into the fail message here?

File: src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    @Test
+    public void testRetryDueToExceptionOnNetworkLayer() throws ExecutionException, InterruptedException, IOException {
+        int numDocs = scaledRandomIntBetween(100, 1000);
+        NodesStatsResponse nodeStats = client().admin().cluster().prepareNodesStats().get();
+        List<NodeStats> dataNodeStats = new ArrayList<>();
+        for (NodeStats stat : nodeStats.getNodes()) {
+            dataNodeStats.add(stat);
+        }
+
+        Collections.shuffle(dataNodeStats, getRandom());
```
Comment (s1monw) [inline src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java]: you can simplify this a bit here:

```Java

NodeStats unluckyNode = randomFrom(Iterables.toArray(nodestats.getNodes()));
```

File: src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java (hunk 2 above)
Comment (s1monw) [inline src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java]: can you make the `exceptionThrown` and AtomicBoolean and define it inside the test such that it's scope is clear?

**Comment by s1monw** [comment]
left some comments - Looks good though can you revert the second commit I think it's good to have it

File: src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java (hunk 1 above)
Comment (brwe) [inline src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java]: Ok, will make a separate commit to remove everywhere.

**Comment by brwe** [comment]
Conflicts are now suppressed (option 2) and I addressed all review comments. Want to take another look?

File: src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 6]
```diff
@@ -73,6 +73,10 @@
     private final static String OP_TYPE_UPDATE = "update";
     private final static String OP_TYPE_DELETE = "delete";
 
+    public static String getActionName() {
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: can't you just make the `ACTION_NAME` constant public?

File: src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java [hunk 7]
```diff
@@ -501,6 +501,10 @@ private void innerCreateNoLock(Create create, IndexWriter writer, long currentVe
                 // #7142: the primary already determined it's OK to index this document, and we confirmed above that the version doesn't
                 // conflict, so we must also update here on the replica to remain consistent:
                 doUpdate = true;
+            } else if (create.origin() == Operation.Origin.PRIMARY && create.autoGeneratedId() && create.canHaveDuplicates() && currentVersion == 1 && create.version() == Versions.MATCH_ANY) {
+                doUpdate = true;
+                updatedVersion = 1;
+
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java]: extra newline.... I think this optimization deserves a rather longish comment. can you add it?

File: src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java [hunk 8]
```diff
@@ -501,6 +501,10 @@ private void innerCreateNoLock(Create create, IndexWriter writer, long currentVe
                 // #7142: the primary already determined it's OK to index this document, and we confirmed above that the version doesn't
                 // conflict, so we must also update here on the replica to remain consistent:
                 doUpdate = true;
+            } else if (create.origin() == Operation.Origin.PRIMARY && create.autoGeneratedId() && create.canHaveDuplicates() && currentVersion == 1 && create.version() == Versions.MATCH_ANY) {
+                doUpdate = true;
```
Comment (kimchy) [inline src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java]: can you add a comment on why we have this check?

**Comment by brwe** [comment]
added comment in InternalEngine. Is it clear enough?

**Comment by s1monw** [comment]
LGTM

**Comment by brwe** [comment]
closed by f45e6ae
